File: OJ practices/LeetCode题目集/19. Remove Nth Node From End of List(medium).java
/*
 * @Author: SourDumplings
 * @Date: 2020-01-19 09:34:11
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/remove-nth-node-from-end-of-list/
 */

/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode(int x) { val = x; }
* }
*/
class Solution
{
    public ListNode removeNthFromEnd(ListNode head, int n)
    {
        ListNode f = head;
        ListNode p = f;
        ListNode t = p;

        for (int i = 0; i < n - 1; i++)
        {
            t = t.next;
        }
        if (t.next == null)
        {
            return f.next;
        }
        else
        {
            while (t.next != null)
            {
                f = p;
                p = p.next;
                t = t.next;
            }
        }

        f.next = p.next;
        return head;
    }
}

File: OJ practices/LeetCode题目集/19. Remove Nth Node From End of List(medium).cpp
/**
 * @file 19. Remove Nth Node From End of List(medium).cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief
 * https://leetcode.cn/problems/remove-nth-node-from-end-of-list/?envType=study-plan-v2&envId=top-interview-150
 * @version 1.0.0
 * @date 2024-11-27
 *
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 * 双指针法，两个指针距离 n 即可
 *
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
	ListNode* removeNthFromEnd(ListNode* head, int n)
	{
		ListNode* tempHead = new ListNode(0, head);
		ListNode* l		   = tempHead;
		ListNode* r		   = head;

		int count = 1;
		while (r->next)
		{
			if (count < n) { ++count; }
			else { l = l->next; }
			r = r->next;
		}
		l->next		  = l->next->next;
		ListNode* ret = tempHead->next;
		delete tempHead;
		return ret;
	}
};